lib/collection/src/wal.rs

```diff
@@ -110,8 +110,35 @@ impl<'s, R: DeserializeOwned + Serialize + Debug> SerdeWal<R> {
             .map_err(|err| WalError::WriteWalError(format!("{err:?}")))
     }
 
-    pub fn read_all(&'s self) -> impl Iterator<Item = (u64, R)> + 's {
-        self.read(self.first_index())
+    pub fn read_all(
+        &self,
+        with_acknowledged: bool,
+    ) -> impl DoubleEndedIterator<Item = (u64, R)> + '_ {
+        if with_acknowledged {
+            self.read(self.first_closed_index())
+        } else {
+            self.read(self.first_index())
+        }
+    }
+
+    pub fn read(&self, from: u64) -> impl DoubleEndedIterator<Item = (u64, R)> + '_ {
+        // We have to explicitly do `from..self.first_index() + self.len(false)`, instead of more
+        // concise `from..=self.last_index()`, because if the WAL is empty, `Wal::last_index`
+        // returns `Wal::first_index`, so we end up with `1..=1` instead of an empty range. 😕
+        //
+        // TODO: Refactor `SerdeWal::last_index`/`Wal::last_index` to return `Option<u64>`? 🤔
```

Comment:
:+1:

Reply:
I'd just remove the `TODO` then. Don't feel much like fixing it now, especially, because then `SerdeWall::last_index` and `Wal::last_index` will behave different.